device/rg28xx/input/input.sh

```diff
@@ -1,361 +1,162 @@
-			;;
-		$RELEASE_L2)
-			if [ $STATE_L2 -eq 1 ]; then
-				KEY_COMBO=0
-				STATE_L2=0
-			fi
-			;;
-		$PRESS_R2)
-			STATE_R2=1
-			COUNT_R2=$((COUNT_R2 + 1))
-			;;
-		$RELEASE_R2)
-			if [ $STATE_R2 -eq 1 ]; then
-				KEY_COMBO=0
-				STATE_R2=0
-			fi
-			;;
-		$PRESS_VOL_UP)
-			STATE_VOL_UP=1
-			COUNT_VOL_UP=$((COUNT_VOL_UP + 1))
-			;;
-		$RELEASE_VOL_UP)
-			if [ $STATE_VOL_UP -eq 1 ]; then
-				KEY_COMBO=0
-				STATE_VOL_UP=0
-			fi
-			;;
-		$PRESS_VOL_DOWN)
-			STATE_VOL_DOWN=1
-			COUNT_VOL_DOWN=$((COUNT_VOL_DOWN + 1))
-			;;
-		$RELEASE_VOL_DOWN)
-			if [ $STATE_VOL_DOWN -eq 1 ]; then
-				KEY_COMBO=0
-				STATE_VOL_DOWN=0
-			fi
-			;;
-		$PRESS_POWER_SHORT)
-			STATE_POWER_SHORT=1
-			COUNT_POWER_SHORT=$((COUNT_POWER_SHORT + 1))
-			;;
-		$RELEASE_POWER_SHORT)
-			if [ $STATE_POWER_LONG -eq 1 ]; then
-				KEY_COMBO=0
-				STATE_POWER_SHORT=0
-				STATE_POWER_LONG=0
+		# Instant Shutdown:
+		2) HALT_SYSTEM sleep poweroff ;;
+		# Sleep XXs + Shutdown:
+		*)
+			if [ ! -e "$POWER_LONG_FILE" ] || [ "$(cat "$POWER_LONG_FILE")" = off ]; then
+				echo on >"$POWER_LONG_FILE"
 			else
-				KEY_COMBO=0
-				STATE_POWER_SHORT=0
-				STATE_POWER_LONG=0
-				if [ "$(GET_VAR "global" "boot/factory_reset")" -eq 0 ]; then
-					FG_PROC_VAL=$(GET_VAR "system" "foreground_process")
-					DPAD_VAL=$(cat "$DPAD")
-					case "$FG_PROC_VAL" in
-						mux*) ;;
-						*)
-							if [ "$STATE_MENU_LONG" -ne 1 ]; then
-								if [ "$DPAD_VAL" -eq 0 ]; then
-									echo 2 >$DPAD
-									MOTO_BUZZ
-								elif [ "$DPAD_VAL" -eq 2 ]; then
-									echo 0 >$DPAD
-									MOTO_BUZZ
-									sleep 0.1
-									MOTO_BUZZ
-								fi
-							fi
-							;;
-					esac
-				fi
+				echo off >"$POWER_LONG_FILE"
 			fi
 			;;
-		$PRESS_POWER_LONG)
-			if [ $STATE_POWER_LONG -eq 0 ]; then
-				STATE_POWER_LONG=1
-				COUNT_POWER_LONG=$((COUNT_POWER_LONG + 1))
-			fi
+	esac
+}
+
+DPAD_TOGGLE() {
+	case "$(GET_VAR system foreground_process)" in
+		mux*) ;;
+		*)
+			case "$(cat "$DPAD_FILE")" in
+				0)
+					echo 2 >"$DPAD_FILE"
+					RUMBLE "$(GET_VAR device board/rumble)" .1
+					;;
+				2)
+					echo 0 >"$DPAD_FILE"
+					RUMBLE "$(GET_VAR device board/rumble)" .1
+					sleep .1
+					RUMBLE "$(GET_VAR device board/rumble)" .1
+					;;
+			esac
 			;;
 	esac
-done &
+}
+
+mkdir -p /tmp/trigger
+echo awake >"$SLEEP_STATE_FILE"
+
+# Start background power listener and sleep timer.
+if [ "$(GET_VAR global boot/factory_reset)" -eq 0 ]; then
+	/opt/muos/device/current/input/trigger/power.sh &
+	/opt/muos/device/current/input/trigger/sleep.sh &
+fi
+
+MONITOR_FG_PROC &
+
+READ_HOTKEYS | while read -r HOTKEY; do
+	# Don't respond to any hotkeys while in charge mode.
+	if [ "$(GET_VAR system foreground_process)" = muxcharge ]; then
+		continue
+	fi
+
+	# During soft sleep, only respond to the power button (to wake back up).
+	if [ "$(cat "$SLEEP_STATE_FILE")" != awake ] && [ "$HOTKEY" != SLEEP ]; then
+		continue
+	fi
+
+	HANDLE_HOTKEY "$HOTKEY"
+done
```

Comment:
We can either have a `&` symbol after the `done` here or place it on the line where this script is called which would be `startup.sh` in either the system or device specific scripts.  Or do both as it can't hurt?

Reply:
Turns out the device startup scripts already run `input.sh` in background, it's just the factory reset flow that didn't. Added `&` in `startup.sh`. (Left it out here as discussed so it's possibly to manually run this script in the foreground when testing.)